hal/src/sercom/v2/uart/mod.rs

```diff
@@ -0,0 +1,1531 @@
+```
+
+[`Buffer`]: crate::dmac::transfer::Buffer
+[`send_with_dma`]: Uart::send_with_dma
+[`receive_with_dma`]: Uart::receive_with_dma
+[`dmac::Transfer`]: crate::dmac::Transfer
+[`Channel`]: crate::dmac::channel::Channel
+[`dmac`]: crate::dmac
+
+"
+)]
+
+mod reg;
+use reg::Registers;
+
+use super::*;
+use core::{
+    convert::{TryFrom, TryInto},
+    marker::PhantomData,
+};
+
+use crate::{
+    target_device as pac,
+    time::Hertz,
+    typelevel::{Is, Sealed},
+};
+use bitflags::bitflags;
+
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+pub use crate::thumbv6m::sercom::v2::uart::*;
+
+#[cfg(feature = "min-samd51g")]
+pub use crate::thumbv7em::sercom::v2::uart::*;
+
+use embedded_hal::{
+    blocking,
+    serial::{Read, Write},
+};
+use nb::Error::WouldBlock;
+use num_traits::{AsPrimitive, PrimInt};
+
+//=============================================================================
+// Character size
+//=============================================================================
+
+/// Size of the SERCOM's `DATA` register
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+pub type DataSize = u16;
+
+/// Size of the SERCOM's `DATA` register
+#[cfg(any(feature = "min-samd51g"))]
+pub type DataSize = u32;
+
+/// Type-level `enum` representing the UART character size
+///
+/// The UART character size affects the word size for the embedded HAL traits.
+/// Eight or less bit transactions use a `u8` word, while nine-bit transactions
+/// use a `u16` word.
+pub trait CharSize: Sealed {
+    /// Word size for the character size
+    type Word: 'static + PrimInt + AsPrimitive<DataSize>;
+
+    /// Bits to write into the `LENGTH` register
+    const BITS: u8;
+}
+
+/// Type alias to recover the `Word` type from an implementation of [`CharSize`]
+pub type Word<C> = <C as CharSize>::Word;
+
+/// [`CharSize`] variant for 5-bit transactions
+pub enum FiveBit {}
+
+/// [`CharSize`] variant for 6-bit transactions
+pub enum SixBit {}
+
+/// [`CharSize`] variant for 7-bit transactions
+pub enum SevenBit {}
+
+/// [`CharSize`] variant for 8-bit transactions
+pub enum EightBit {}
+
+/// [`CharSize`] variant for 9-bit transactions
+pub enum NineBit {}
+
+impl Sealed for FiveBit {}
+impl CharSize for FiveBit {
+    type Word = u8;
+    const BITS: u8 = 0x5;
+}
+
+impl Sealed for SixBit {}
+impl CharSize for SixBit {
+    type Word = u8;
+    const BITS: u8 = 0x6;
+}
+
+impl Sealed for SevenBit {}
+impl CharSize for SevenBit {
+    type Word = u8;
+    const BITS: u8 = 0x7;
+}
+
+impl Sealed for EightBit {}
+impl CharSize for EightBit {
+    type Word = u8;
+    const BITS: u8 = 0x0;
+}
+
+impl Sealed for NineBit {}
+impl CharSize for NineBit {
+    type Word = u16;
+    const BITS: u8 = 0x1;
+}
+
+//=============================================================================
+// Stop bits, parity, baud rate
+//=============================================================================
+
+/// Number of stop bits in a UART frame
+pub enum StopBits {
```

Comment:
To implement that USB-UART, I had to add `#[derive(Copy, Clone)]` to `StopBits` and `Parity`.  It also might be nice to have a way to read the current settings out of a Uart if that's not too much trouble to implement.